### PR TITLE
KinKal fit extension

### PR DESCRIPTION
**Pull request by brownd1978**
Factorize common fit functions out of LoopHelixFit into KKFit, to facilitate implementing other kinds of fits.  Implement fit extension; ie hit and cluster adding and refit.  Remove internal references to HelixSeed from KalSeed, and replace with an external Assns, to support other kinds of fit seeds.  Add an explicit KKTrack subclass of KinKal::Track to hold mu2e-specific hit and material payload.   Also add a mu2e-specific KKStraw subclass of KinKal::Straw to allow storing strawId, needed to define material uniqueness.

### CODE REVIEW

File: JobConfig/reco/prolog.fcl [hunk 1]
```diff
@@ -100,7 +100,8 @@ Reconstruction : {
 	HelixSeedCollections : [ "MHDeM" ]
         ComboHitCollection : "makeSH"
         StrawHitFlagCollection : "FlagBkgHitss:StrawHits"
```
Comment (gianipez) [inline JobConfig/reco/prolog.fcl:102]: `FlagBkgHitss` is this a typo? 2 's'?

Comment (brownd1978) [reply JobConfig/reco/prolog.fcl:102]: good catch!  Not sure why, but this worked.  I'll investigate.

**Review by AndrewEdmonds11 — APPROVED** [review]
The TrkAna-related changes look good to me. I like that InfoStructHelper::fillHelixInfo now takes a HelixSeed rather than a KalSeed - much more sensible!

File: JobConfig/reco/prolog.fcl (hunk 1 above)
Comment (brownd1978) [reply JobConfig/reco/prolog.fcl]: This collection isn't currently being used so the config error was irrelevant.  I fixed it anyways.

**Comment by brownd1978** [comment]
@FNALbuild run build test

**Review by sophiemiddleton — APPROVED** [review]
All looks good to me

File: Mu2eKinKal/inc/KKFitSettings.hh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#include "DataProducts/inc/PDGCode.hh"
+#include "RecoDataProducts/inc/TrkFitDirection.hh"
+namespace mu2e {
+  namespace Mu2eKinKal{
+    using Name    = fhicl::Name;
+    using Comment = fhicl::Comment;
+
+    // struct for defining the KinKal Config object
+    struct KKConfig {
+      fhicl::Atom<int> maxniter { Name("MaxNIter"), Comment("Maximum number of algebraic iteration steps in each fit meta-iteration"), 10 };
```
Comment (gianipez) [inline Mu2eKinKal/inc/KKFitSettings.hh:20]: are all these default necessary? or they could live in the corresponding prolog file?

File: RecoDataProducts/inc/KalSeed.hh [hunk 3]
```diff
@@ -40,12 +40,12 @@ namespace mu2e {
     UInt_t nTrajSegments() const { return _nseg; }
     bool hasCaloCluster() const { return _chit.caloCluster().isNonnull(); }
     art::Ptr<CaloCluster> const& caloCluster() const { return _chit.caloCluster(); }
-    art::Ptr<HelixSeed> const& helix() const { return _helix; }
```
Comment (gianipez) [inline RecoDataProducts/inc/KalSeed.hh:43]: I understand why you removed it; users where using it improperly. BUT I have thinking quite a bit about it and I remembered its real need; I'm using it to associate a reco track with the corresponding Helix that started it. 
Looking at the KalSeedFit_module I don't see a parallel structure of HelixSeedCollection pushed into the event.
Is there a different way I'm not aware of to match a track with the seed Helix?

File: RecoDataProducts/inc/KalSeedAssns.hh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#include "RecoDataProducts/inc/KalSeed.hh"
+#include "RecoDataProducts/inc/HelixSeed.hh"
+#include "canvas/Persistency/Common/Assns.h"
+namespace mu2e {
+  // Assns between a KalSeed and the HelixSeed it was created from
+  typedef art::Assns<KalSeed,HelixSeed> KalHelixAssns;
```
Comment (gianipez) [inline RecoDataProducts/inc/KalSeedAssns.hh:6]: here it is! please, ignore my previous comment.

**Review by gianipez — CHANGES_REQUESTED** [review]
Hi Dave,
My last comment is only about those default values in the Config struct. Other than that, the PR looks good to me!

**Comment by brownd1978** [comment]
@FNALbuild run build test

**Comment by brownd1978** [comment]
I moved the parameter defaults all into prolog.

**Review by gianipez — APPROVED** [review]
It looks great now!

**Pull request merged**